goodtables/compat.py

```diff
@@ -30,22 +31,22 @@
     basestring = basestring
     numeric_types = (int, long, float)
 
-    def csv_reader(data, dialect=csv.excel, **kwargs):
+    def csv_reader(data, dialect=csv.excel, header_index=0, **kwargs):
         """Read text stream (unicode on Py2.7) as CSV."""
 
-        first_lines = list(islice(data, 5))
-        try:
-            dialect = csv.Sniffer().sniff(''.join(first_lines))
-            dialect.delimiter = dialect.delimiter.encode('utf-8')
-            dialect.quotechar = dialect.quotechar.encode('utf-8')
-        except csv.Error:
-            dialect = csv.excel
-
+        first_lines = list(islice(data, header_index, header_index + 2))
+        dialect = detect_dialect(first_lines)
+        
+        while not re.findall('[^\w ]', dialect.delimiter): 
```

Comment:
That leaves uncovered the case for multiple spaces (ex: a tab that is actually 4 spaces). I think it's better to have all cases covered than to be very fast in this case because that check won't execute many times.